programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -1128,29 +1359,40 @@ fn update_caller_account(
     *caller_account.lamports = callee_account.get_lamports();
     *caller_account.owner = *callee_account.get_owner();
 
-    if direct_mapping && caller_account.original_data_len > 0 {
-        // Since each instruction account is directly mapped in a memory region
-        // with a *fixed* length, upon returning from CPI we must ensure that the
-        // current capacity is at least the original length (what is mapped in
-        // memory), so that the account's memory region never points to an
-        // invalid address.
-        let min_capacity = caller_account.original_data_len;
-        if callee_account.capacity() < min_capacity {
-            callee_account.reserve(min_capacity.saturating_sub(callee_account.capacity()))?
-        }
+    let mut zero_all_spare_capacity = false;
+    if direct_mapping {
+        if let Some(region) = account_data_region(
+            memory_mapping,
+            caller_account.vm_data_addr,
+            caller_account.original_data_len,
+        )? {
+            // Since each instruction account is directly mapped in a memory region
+            // with a *fixed* length, upon returning from CPI we must ensure that the
+            // current capacity is at least the original length (what is mapped in
+            // memory), so that the account's memory region never points to an
+            // invalid address.
+            let min_capacity = caller_account.original_data_len;
+            if callee_account.capacity() < min_capacity {
+                callee_account.reserve(min_capacity.saturating_sub(callee_account.capacity()))?;
+                zero_all_spare_capacity = true;
+            }
 
-        // If an account's data pointer has changed - because of CoW or because
-        // of using AccountSharedData directly (deprecated) - we must update the
-        // corresponding MemoryRegion in the caller's address space. Address
-        // spaces are fixed so we don't need to update the MemoryRegion's length.
-        let region = memory_mapping.region(AccessType::Load, caller_account.vm_data_addr)?;
-        let callee_ptr = callee_account.get_data().as_ptr() as u64;
-        if region.host_addr.get() != callee_ptr {
-            region.host_addr.set(callee_ptr);
+            // If an account's data pointer has changed - because of CoW or because
+            // of using AccountSharedData directly (deprecated) - we must update the
```

Comment:
nit: how about mentioning the above `.reserve()` call as well for completeness?

```rust
    // If an account's data pointer has changed - because of CoW or because
    // of using AccountSharedData directly (deprecated) or just above `.reserve()` 
    // - we must update the
```

Reply:
updated